pkg/pillar/types/global.go

```diff
@@ -103,312 +110,367 @@ func (gs *GlobalStatus) UpdateItemValuesFromGlobalConfig(gc GlobalConfig) {
+}
+
+func (specMap *ConfigItemSpecMap) ParseItem(configMap *ConfigItemValueMap, item config.ConfigItem) error {
+	key := item.Key
+	configItem := ConfigItemValue{
+		Key: key,
+	}
+	agentSetting := false
+	globalSetting := false
+	configSpecMap := specMap.GlobalSettings
+	if _, ok := specMap.GlobalSettings[key]; ok {
+		configSpecMap = specMap.GlobalSettings
+		globalSetting = true
+	} else {
+		components := strings.Split(key, ".")
+		if _, ok := specMap.AgentSettings[components[1]]; ok && (components[0] == "debug" || components[0] == "agent") {
+			configSpecMap = specMap.AgentSettings[components[1]]
+			configItem.Key = strings.Join(components[2:], "")
+			key = components[1]
+			agentSetting = true
+		} else {
+			goto notfound
+		}
+	}
+
+	if configSpecMap[key].ItemType == CONFIGITEM_TYPE_INT {
+		i64, err := strconv.ParseUint(item.Value, 10, 32)
+		if err == nil {
+			configItem.IntValue = uint32(i64)
+		} else {
+			configItem.IntValue = specMap.GlobalSettings[key].IntDefault
+			log.Errorf("Failed to parse int value for key %s, value: %s err: %s", key, item.Value, err)
+			return err
+		}
+		if configItem.IntValue > specMap.GlobalSettings[key].IntMax {
+			configItem.IntValue = specMap.GlobalSettings[key].IntMax
+			log.Errorf("Key %s larger than max was changed to max %d", key, specMap.GlobalSettings[key].IntMax)
+		} else if configItem.IntValue < specMap.GlobalSettings[key].IntMin {
+			configItem.IntValue = specMap.GlobalSettings[key].IntMin
+			log.Errorf("Key %s smaller than min was changed to min %d", key, specMap.GlobalSettings[key].IntMax)
+		}
+	} else if configSpecMap[key].ItemType == CONFIGITEM_TYPE_TRISTATE {
+		newTs, err := ParseTriState(item.Value)
+		if err == nil {
+			configItem.TriStateValue = newTs
+		} else {
+			log.Errorf("Failed to parse tristate value for key %s, value: %s err: %s", key, item.Value, err)
+			return err
+		}
+	} else if configSpecMap[key].ItemType == CONFIGITEM_TYPE_BOOL {
+		newBool, err := strconv.ParseBool(item.Value)
+		if err == nil {
+			configItem.BoolValue = newBool
+		} else {
+			log.Errorf("Failed to parse bool value for key %s, value: %s err: %s", key, item.Value, err)
+			return err
+		}
+	} else if configSpecMap[key].ItemType == CONFIGITEM_TYPE_STRING {
+		configItem.StringValue = item.Value
+	}
+notfound:
+	if globalSetting {
+		configMap.GlobalSettings[key] = configItem
+	} else if agentSetting {
+		components := strings.Split(item.Key, ".")
+		if len(components) == 3 && components[2] == "loglevel" {
+			configMap.AgentSettings[key]["loglevel"] = configItem
+		} else if len(components) == 4 && components[2] == "remote" && components[3] == "loglevel" {
+			configMap.AgentSettings[key]["remote.loglevel"] = configItem
+		}
+	} else {
+		err := fmt.Errorf("Unknown configItem %s value %s\n", key, item.Value)
+		log.Errorf("parseConfigItems: %s", err)
+		return err
+	}
+	return nil
+}
+
+func (valueMap ConfigItemValueMap) FormatConfig(specMap *ConfigItemSpecMap) ConfigItemValueMap {
```

Comment:
This function is supposed to change the values in ConfigItemValueMap - isn't it? Then shouldn't valueMap is a pointer to ConfigItemValueMap?